File: solutions/typescript/0017-letter-combinations-of-a-phone-number.ts
function letterCombinations(digits: string): string[] {
    if (digits.length === 0) return [];

    const phoneMap: string[] = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"];
    const output: string[] = [];

    function backtrack(combination: string, nextDigits: string) {
        if (nextDigits.length === 0) {
            output.push(combination);
        } else {
            const letters: string = phoneMap[parseInt(nextDigits[0]) - 2];
            for (const letter of letters) {
                backtrack(combination + letter, nextDigits.slice(1));
            }
        }
    }

    backtrack("", digits);
    return output;
}

File: solutions/go/0017-letter-combinations-of-a-phone-number.go
func letterCombinations(digits string) []string {
	if digits == "" {
		return []string{}
	}

	phoneMap := []string{"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"}
	var output []string

	var backtrack func(combination string, nextDigits string)
	backtrack = func(combination string, nextDigits string) {
		if nextDigits == "" {
			output = append(output, combination)
		} else {
			letters := phoneMap[nextDigits[0]-'2']
			for _, letter := range letters {
				backtrack(combination+string(letter), nextDigits[1:])
			}
		}
	}

	backtrack("", digits)
	return output
}
